qcodes/interactive_widget.py

```diff
@@ -0,0 +1,436 @@
+"""This file contains functions to displays an interactive widget
+with information about `qcodes.experiments()`."""
+
+import io
+import math
+import operator
+from functools import partial, reduce
+from typing import Any, Callable, Dict, Optional, Sequence, Tuple
+
+import matplotlib.pyplot as plt
+from IPython.core.display import display
+from IPython.display import clear_output
+from ipywidgets import (
+    HTML,
+    Box,
+    Button,
+    GridspecLayout,
+    Label,
+    Layout,
+    Output,
+    Tab,
+    Textarea,
+    VBox,
+)
+from ruamel.yaml import YAML
+
+import qcodes
+from qcodes.dataset import initialise_or_create_database_at
+from qcodes.dataset.data_export import get_data_by_id
+from qcodes.dataset.data_set import DataSet
+from qcodes.dataset.plotting import plot_dataset
+
+_META_DATA_KEY = "widget_notes"
+
+
+def _get_in(nested_keys: Sequence[str], dct: Dict) -> Dict:
+    """ Returns dct[i0][i1]...[iX] where [i0, i1, ..., iX]==nested_keys."""
+    return reduce(operator.getitem, nested_keys, dct)
+
+
+def button(
+    description: str,
+    button_style: Optional[str] = None,
+    on_click: Optional[Callable[[Any], None]] = None,
+    tooltip: Optional[str] = None,
+    layout_kwargs: Optional[Dict[str, Any]] = None,
+    button_kwargs: Optional[Dict[str, Any]] = None,
+) -> Button:
+    """Returns a `ipywidgets.Button`."""
+    layout_kwargs = layout_kwargs or {}
+    but = Button(
+        description=description,
+        button_style=button_style,
+        layout=Layout(
+            height=layout_kwargs.pop("height", "auto"),
+            width=layout_kwargs.pop("width", "auto"),
+            **layout_kwargs,
+        ),
+        tooltip=tooltip or description,
+        **(button_kwargs or {}),
+    )
+    if on_click is not None:
+        but.on_click(on_click)
+    return but
+
+
+def label(description: str) -> Label:
+    """Returns a `ipywidgets.Label` with text."""
+    return Label(
+        value=description, layout=Layout(height="max-content", width="auto")
+    )
+
+
+def _update_nested_dict_browser(
+    nested_keys: Sequence[str], table: Dict[Any, Any], box: Box
+) -> Callable[[Button], None]:
+    def update_box(_: Button) -> None:
+        box.children = (_nested_dict_browser(nested_keys, table, box),)
+
+    return update_box
+
+
+def _nested_dict_browser(
+    nested_keys: Sequence[str],
+    table: Dict[Any, Any],
+    box: Box,
+    max_nrows: int = 30,
+) -> GridspecLayout:
+    """Generates a `GridspecLayout` of the ``nested_keys`` in ``table`` which is
```

Comment:
i guess `table` is the same as `dict` in `_nested_dict_browser`? some uniformity with naming would help :)

Reply:
Fixed in 23b6b3523. I renamed everything to `nested_dict` because it's more obvious what that is.